File: Kalorie/Kalorie/Core/UseCases/SignOutUseCase.swift
//
//  SignOutUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 07.08.2026.
//

import Foundation

protocol SignOutUseCaseProtocol {
    func callAsFunction() throws
}

struct SignOutUseCase: SignOutUseCaseProtocol {

    // MARK: - Properties

    private let authCommandProvider: any AuthCommandProviderProtocol
    private let snapshotStore: any PendingMergeSnapshotStoreProtocol
    private let googleSessionProvider: any GoogleSessionProviderProtocol

    // MARK: - Init

    init(
        authCommandProvider: any AuthCommandProviderProtocol,
        snapshotStore: any PendingMergeSnapshotStoreProtocol,
        googleSessionProvider: any GoogleSessionProviderProtocol
    ) {
        self.authCommandProvider = authCommandProvider
        self.snapshotStore = snapshotStore
        self.googleSessionProvider = googleSessionProvider
    }

    // MARK: - Functions

    // The snapshot is dropped before signing out: it is only ever valid for the account the
    // merge was destined for, and signing out immediately provisions a fresh anonymous uid.
    // A snapshot surviving into that new account would be resumed against it on the next
    // launch and write the previous user's food there.
    func callAsFunction() throws {
        try snapshotStore.delete()
        try authCommandProvider.signOut()
        googleSessionProvider.clearSession()
    }
}

#if DEBUG
struct SignOutUseCaseFake: SignOutUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction() throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SignOutUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthCommandProviderProtocol
import antoni.kalorie.core.auth.GoogleSessionProviderProtocol
import antoni.kalorie.core.auth.PendingMergeSnapshotStoreProtocol

interface SignOutUseCaseProtocol {
    suspend operator fun invoke()
}

class SignOutUseCase(
    private val authCommandProvider: AuthCommandProviderProtocol,
    private val snapshotStore: PendingMergeSnapshotStoreProtocol,
    private val googleSessionProvider: GoogleSessionProviderProtocol,
) : SignOutUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke() {
        snapshotStore.delete()
        authCommandProvider.signOut()
        googleSessionProvider.clearSession()
    }
}
